### PR TITLE
Addon-Docs: Upgrade to MDX3

### LINKED ISSUE #25182: Upgrade to MDX 3
## Breaking Changes

> The main breaking change is that Node.js 16 is now the minimum supported version. 
Across the ecosystem there were several small internal breaking changes. Everything’s released already. You can update all plugins now. If you ran into problems before, it should work now.
We also removed some infrequently used deprecated APIs. You’re likely fine but gloss over the [v3 migration guide](https://mdxjs.com/migrating/v3/) if you get errors.
Important to note when you use your lesser-known but powerful evaluate, run, or outputFormat: 'function-body' APIs, please pass the baseUrl option. That makes sure import.meta.url, import, and export work. You’ll get a runtime error when those features are used otherwise.

from https://mdxjs.com/blog/v3/

## Migration Guide

https://mdxjs.com/migrating/v3/

## Tasks

```[tasklist]
### Tasks
- [x] Update to MDX 3
- [ ] Write automigrations/codemods, if necessary
- [ ] Update documentation
- [x] Upgrade MDX in @storybook/docs-mdx
- [x] Deprecate/remove @storybook/mdx1-csf and @storybook/mdx2-csf packages
- [x] Refactor addon-docs to use mdx 3 compiler directly
```

### PR DESCRIPTION
**Pull request by yannbf**
Closes #25182 

## What I did

- Remove the jest-transform-mdx import used for storyshots
- Remove support for CSF in MDX
- Remove usage of @storybook/mdx1-csf and @storybook/mdx2-csf
- Bring the compiler code from @storybook/mdx2-csf into addon docs, alongside with @mdx-js/mdx
- Use MDX 3
- Upgrade @storybook/docs-mdx to a version that uses MDX3

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access docs stories, they should work
4. Build Storybook, it should work
5. Serve the built Storybook and check the stories, just in case

### Documentation

- [x] Add or update documentation reflecting your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-25303-sha-7c4e18ec`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25303-sha-7c4e18ec). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-25303-sha-7c4e18ec`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25303-sha-7c4e18ec) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`yann/upgrade-mdx`](https://github.com/storybookjs/storybook/tree/yann/upgrade-mdx) |
| **Commit** | [`7c4e18ec`](https://github.com/storybookjs/storybook/commit/7c4e18ec1cbc2b8bb58e93733dd42dc95eedcc92) |
| **Datetime** | Tue Jan  2 12:53:00 UTC 2024 (`1704199980`) |
| **Workflow run** | [7385987242](https://github.com/storybookjs/storybook/actions/runs/7385987242) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=25303`_
</details>

### CODE REVIEW

**Review by shilman — CHANGES_REQUESTED** [review]
Can we also get rid of `mdx1-csf` and `mdx2-csf` while we're at it?

File: code/addons/docs/package.json [hunk 1]
```diff
@@ -98,15 +98,14 @@
     "prep": "node --loader ../../../scripts/node_modules/esbuild-register/loader.js -r ../../../scripts/node_modules/esbuild-register/register.js ../../../scripts/prepare/bundle.ts"
   },
   "dependencies": {
-    "@jest/transform": "^29.3.1",
-    "@mdx-js/react": "^2.1.5",
+    "@babel/core": "^7.12.3",
```
Comment (valentinpalkovic) [inline code/addons/docs/package.json:101]: Do you have any substantial reason why not the latest version of @babel/core is used?

File: code/addons/docs/package.json [hunk 2]
```diff
@@ -120,6 +119,7 @@
     "ts-dedent": "^2.0.0"
   },
   "devDependencies": {
+    "@mdx-js/mdx": "^3.0.0",
```
Comment (valentinpalkovic) [inline code/addons/docs/package.json:122]: Can you elaborate on why prebundling is necessary here?

File: code/addons/docs/src/mdx-loader.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  let result: string;
+  try {
+    result = await compile(content, options);
+  } catch (err: any) {
+    console.error('Error loading:', this.resourcePath);
+    return callback(err);
+  }
+
+  const code = `${DEFAULT_RENDERER}\n${result}`;
+  return callback(null, code);
```
Comment (valentinpalkovic) [inline code/addons/docs/src/mdx-loader.ts:42]: This is confusing.

```suggestion
  try {
    const result = await compile(content, options);
    const code = `${DEFAULT_RENDERER}\n${result}`;
    return callback(null, code);
  } catch (err: any) {
    console.error('Error loading:', this.resourcePath);
    return callback(err);
  }
```

File: code/lib/core-server/package.json [hunk 4]
```diff
@@ -68,7 +68,7 @@
     "@storybook/core-events": "workspace:*",
     "@storybook/csf": "^0.1.2",
     "@storybook/csf-tools": "workspace:*",
-    "@storybook/docs-mdx": "^0.1.0",
+    "@storybook/docs-mdx": "0.1.1--canary.13.929b700.0",
```
Comment (valentinpalkovic) [inline code/lib/core-server/package.json:71]: Reminder to use a proper version here.

File: code/addons/docs/package.json (hunk 2 above)
Comment (yannbf) [reply code/addons/docs/package.json:122]: because of ESM only issues. The prebundling generates CJS compatible code

File: code/addons/docs/package.json (hunk 1 above)
Comment (yannbf) [reply code/addons/docs/package.json:101]: This was just the version defined in mdx2-csf. The caret should resolve to the latest version though. I can upgrade as well.

File: code/lib/core-server/src/utils/__tests__/index-extraction.test.ts [hunk 5]
```diff
@@ -604,53 +604,4 @@ describe('docs entries from story extraction', () => {
       }
     `);
   });
-  it(`Only adds a docs entry and not a story entry when an input has the "docsOnly" tag`, async () => {
```
Comment (yannbf) [inline code/lib/core-server/src/utils/__tests__/index-extraction.test.ts:607]: @shilman I need assistance here, I am not sure if we can just remove this test as it was targetting .stories.mdx, or if we need to test it differently with a new .mdx stub

File: code/lib/core-server/src/utils/StoryIndexGenerator.test.ts [hunk 6]
```diff
@@ -292,51 +291,6 @@ describe('StoryIndexGenerator', () => {
       });
     });
 
-    describe('mdx tagged components', () => {
-      it('adds docs entry with docs enabled', async () => {
```
Comment (yannbf) [inline code/lib/core-server/src/utils/StoryIndexGenerator.test.ts:296]: @shilman I need assistance here, I am not sure if we can just remove this test as it was targetting .stories.mdx, or if we need to test it differently with a new .mdx stub

File: MIGRATION.md [hunk 7]
```diff
@@ -380,6 +384,24 @@ To summarize:
...
 
+### MDX related changes
+
+#### MDX is upgraded to v3
+
+Storybook now uses MDX3 under the hood. This change contains many improvements and a few small breaking changes that possibly won't affect you. However we recommend checking the [migration notes from MDX here](https://mdxjs.com/blog/v3/).
+
+#### Dropping support for *.stories.mdx (CSF in MDX) format
+
+_Has automigration_
```
Comment (shilman) [inline MIGRATION.md:395]: ???

Comment (shilman) [reply MIGRATION.md:395]: Has a codemod, not an automigration I believe

File: code/addons/docs/src/compiler/types.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  throwIfNamespace?: false;
+  runtime?: 'classic' | 'automatic';
+  importSource?: string;
+}
+
+export type MdxCompileOptions = Parameters<typeof mdxCompile>[1];
+export type BabelOptions = Parameters<typeof transformAsync>[1];
+
+export interface CompileOptions {
+  skipCsf?: boolean;
```
Comment (shilman) [inline code/addons/docs/src/compiler/types.ts:16]:
```suggestion
```

File: code/addons/docs/src/compiler/types.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  mdxCompileOptions?: MdxCompileOptions;
+  jsxOptions?: JSXOptions;
```
Comment (shilman) [inline code/addons/docs/src/compiler/types.ts:18]: Do we still need this?

File: code/addons/docs/src/preset.ts [hunk 10]
```diff
@@ -27,25 +23,13 @@ const getResolvedReact = async (options: Options) => {
...
-    configureJsx: boolean;
-    /**
-     * @deprecated
-     * Use `jsxOptions` to customize options used by @babel/preset-react
-     */
-    mdxBabelOptions?: any;
-    /** @deprecated */
-    sourceLoaderOptions: any;
     csfPluginOptions: CsfPluginOptions | null;
     jsxOptions?: JSXOptions;
```
Comment (shilman) [inline code/addons/docs/src/preset.ts:34]: do we need this?

File: code/lib/core-server/src/utils/StoryIndexGenerator.test.ts (hunk 6 above)
Comment (shilman) [reply code/lib/core-server/src/utils/StoryIndexGenerator.test.ts:296]: Let's get rid of it

File: code/lib/core-server/src/utils/__tests__/index-extraction.test.ts (hunk 5 above)
Comment (shilman) [reply code/lib/core-server/src/utils/__tests__/index-extraction.test.ts:607]: Get rid of it

File: code/addons/docs/src/preset.ts (hunk 10 above)
Comment (yannbf) [reply code/addons/docs/src/preset.ts]: nope!

File: code/addons/docs/src/compiler/types.ts (hunk 9 above)
Comment (yannbf) [reply code/addons/docs/src/compiler/types.ts]: nope!

File: code/addons/docs/src/compiler/types.ts (hunk 8 above)
Comment (yannbf) [reply code/addons/docs/src/compiler/types.ts]: good catch!

File: MIGRATION.md (hunk 7 above)
Comment (yannbf) [reply MIGRATION.md]: Good catch, I got confused

**Review by shilman — APPROVED** [review]

File: MIGRATION.md [hunk 11]
```diff
@@ -380,6 +384,24 @@ To summarize:
...
+
+Storybook now uses MDX3 under the hood. This change contains many improvements and a few small breaking changes that possibly won't affect you. However we recommend checking the [migration notes from MDX here](https://mdxjs.com/blog/v3/).
+
+#### Dropping support for *.stories.mdx (CSF in MDX) format
+
+In Storybook 7, we deprecated the ability of using MDX both for documentation and for defining stories in the same .stories.mdx file. It is now removed, and Storybook won't support .stories.mdx files anymore. We provide migration scripts to help you onto the new format.
+
+Alongside with this change, the `jsxOptions` configuration was removed as it is not used anymore.
+
+[More info here](https://storybook.js.org/docs/writing-docs/mdx#automigration).
```
Comment (yannbf) [inline MIGRATION.md:399]:
```suggestion
[More info here](https://storybook.js.org/docs/migration-guide#storiesmdx-to-mdxcsf).
```

File: MIGRATION.md [hunk 12]
```diff
@@ -380,6 +384,24 @@ To summarize:
 - This allows users and (test) integrators to run or build storybook without docgen, boosting the user performance and allows tools to give quicker feedback.
 - This will make sure that we can one day lazy load docgen, without changing how stories are rendered.
 
+### MDX related changes
+
+#### MDX is upgraded to v3
+
+Storybook now uses MDX3 under the hood. This change contains many improvements and a few small breaking changes that possibly won't affect you. However we recommend checking the [migration notes from MDX here](https://mdxjs.com/blog/v3/).
```
Comment (yannbf) [inline MIGRATION.md:391]:
```suggestion
Storybook now uses MDX3 under the hood. This change contains many improvements and a few small breaking changes that probably won't affect you. However we recommend checking the [migration notes from MDX here](https://mdxjs.com/blog/v3/).
```

File: code/addons/docs/src/preset.ts [hunk 13]
```diff
@@ -27,47 +23,24 @@ const getResolvedReact = async (options: Options) => {
     // addon-docs, causing addon-docs's dependencies not to be hoisted.
     // This might also affect regular users who have a similar setup.
     // Explicitly alias @mdx-js/react to avoid this issue.
-    mdx: resolvedReact.mdx ?? dirname(require.resolve('@mdx-js/react/package.json')),
+    mdx: resolvedReact.mdx ?? dirname(require.resolve('@mdx-js/react')),
```
Comment (ndelangen) [inline code/addons/docs/src/preset.ts:26]: why did you remove `/package.json` here?

Comment (yannbf) [reply code/addons/docs/src/preset.ts:26]: Because @mdx-js/react v3 has an exports map which does not contain package.json as an entry
